### ISSUE TITLE
Configuration cache fails if TaskOutputs are used as an input to another task

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#23093]
Given the following example project on the latest 8.0 snapshot:

```                                                                                                                                             
tasks {
  val foo by registering {
  }

  val bar by registering {
    inputs.files(foo.map {it.outputs})
  }
}
```

When I run `gradlew bar`, it fails with 

```
Configuration cache state could not be cached: field `queue` of `java.util.WeakHashMap` bean found in field `allValues` of `org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$DefaultCrossBuildInMemoryCache` bean found in field `cache` of `org.gradle.internal.properties.annotations.DefaultTypeMetadataStore` bean found in field `metadataStore` of `org.gradle.internal.properties.bean.RuntimeBeanNodeFactory` bean found in field `nodeFactory` of `org.gradle.internal.properties.bean.DefaultPropertyWalker` bean found in field `propertyWalker` of `org.gradle.api.internal.tasks.DefaultTaskOutputs` bean found in input property `$1` of task `:bar` of type `org.gradle.api.DefaultTask`: error writing value of type 'java.lang.ref.ReferenceQueue'
> Unable to make field private volatile java.lang.ref.Reference java.lang.ref.ReferenceQueue.head accessible: module java.base does not "opens java.lang.ref" to unnamed module @4c87224b
```

**Comment by xaviarias** [comment]
Could you please try without using delegated properties? Seems like a possible duplicate of #23092 otherwise.

**Comment by oehme** [comment]
You're right, this is the same as https://github.com/gradle/gradle/issues/23092

**Comment by bamboo** [comment]
Ah, interesting, I think I see what's going on. Thanks for reopening it, @oehme  👍

**Comment by eskatos** [comment]
Thank you for providing a valid reproducer.

The issue is in the backlog of the relevant team and is prioritized by them.

**Comment by oehme** [comment]
Hey folks, I read in Gradle 8.1 "the configuration cache is now stable and ready for general use". Was this fixed in 8.1?

**Comment by oehme** [comment]
Tried it out right now and the error remains the same. This means the config cache still doesn't work with idiomatic Kotlin build scripts. 

This causes a lot of frustration for my clients - They wanted to be good citizens, so they upgraded their builds to use the lazy configuration API everywhere. Now I have to tell them that because they were good citizens, they can't have the config cache.

**Comment by mlopatkin** [comment]
What trips CC over in this particular scenario is the use of TaskOutputs as an input. Is there any specific reason to prefer `inputs.files(foo.map {it.outputs})` to just `inputs.files(foo)` (this works with CC for some time already) or adding the outputs individually?

**Comment by oehme** [comment]
`it.outputs` was only one example, but I agree that it was probably a bad one :) I had similar issues when using `map {}` with other task properties, but those seem to have been fixed. I agree that `it.outputs` is not worth worrying about.

**Comment by mlopatkin** [comment]
>  I agree that `it.outputs` is not worth worrying about.

I'd rather keep this issue open, so we can do something with `TaskOutputs`, the current failure message is not helpful.
